Add anonymous states

T0mstone opened the pull request:
See #2425 for context.

I don't expect anything out of this PR, I just had some time and wanted to try implementing it, tho I'd also be happy to act on feedback and get this merged.

## Implementation
Most notably is the new type `state.key`:
```typ
#state.key(
  // positional
  key: str | none,
  // default: auto
  public: bool | auto
)
```
The combination `state.key(none, public: true)` is forbidden.
The value for `public: auto` is `true` when `key` is a `str` and `false` when it is `none`.

There's also a method `state.key.as-str` that returns the inner string, but only if the key is public.

The constructor for `state` has been changed to
```typ
#state(
  // positional
  init,
  // optional
  key: str | state.key
)
```
So the previous `state(s, init)` is now `state(init, key: s)` and it is actually just a shorthand for `state(init, key: state.key(s))`.

The new `state(init)` without a key is actually a bit special and not equivalent to `state(init, key: state.key(none))` (see below).

### Anonymous Keys
States with anonymous keys come in three flavors:
- `state(init, key: state.key(none))`: This is the most basic idea of an anonymous key and just uses the span of the `state.key` call as the key.
- `state(init, key: state.key(s, public: false))`: This uses both the span of the `state.key` call and the string `s` as the state key, allowing the use of anonymous keys in loops and such.
- `state(init)`: This is a sort of compromise between the two previous ones: It uses both the span of the `state` call and the hash of `init` as the key. My reasoning for this is that states with the same key and different initial values are rarely if ever useful and this is a sort of compromise to make something like https://github.com/typst/typst/issues/2425#issuecomment-1892599687 work (at least as long as the arguments to `create_state` are different -- two separate calls with the same argument would still create the same state).

## Other Implementation Notes
I'd like it to be `state.key()` instead of `state.key(none)`, but I couldn't find a way to make this work without having to manually parse the arguments, which I'd like to avoid to keep using the auto-generated arg parsing errors.

Also, I haven't written any tests for now, but I'd be happy to do so if we decide to merge this.

T0mstone commented:
> Could you expand on the motivation for the new API surface?

From the discussion in #2425, it seemed to me like `state(init)` with a totally argument-independent key would be too unexpected, so I thought making the argument-independence explicit in `state(init, key: state.key())` would make the API more intuitive.

I'm not 100% satisfied with the way I implemented `state(init)` (i.e. the third point above), since there's still wonky behavior like `(1, 2).map(state)` being two different states but `(1, 1).map(state)` being two copies of the same one, but I don't really have a better idea either. Maybe documenting this well enough and telling people to use string-based private keys if they ever encounter such a collision is good enough. Or maybe this single-argument `state` call should just get removed and we could go back to the previous syntax of only having `state(key, init)`, but still keeping the new feature of private string keys (which is kind of the main feature of this PR).

If we go down that route, maybe the API should just be `state(key: str, init, #[named] public: bool)` instead.

On the other hand, that'd go back to being a pretty small change and wouldn't really address users' confusion over why there has to be a string key in addition to giving it a variable name.

T0mstone commented:
That just gave me another idea: What about doing it like I just described, but also making `key` named and just defaulting to `""`. Most simple usages of `state` are just assigning it to a variable once and for those, the span is unique enough, so the string isn't needed to disambiguate and can just be an arbitrary value (in this case the empty string).

At least I assume that's most usages, maybe the actual data contradicts me and this doesn't make much sense...

cAttte commented:
regarding the other idea of using the state's initial value as a component of the key/hash: it makes sense when we think of the typical "static state" use case, where you have something like `#let my-state = state(5)` and you're done. but the whole point of still allowing string keys is to aid in the "dynamic state" use case, where a user may have something like `(a, b).map(state)`.

in this scenario, a user might naively test their function with a ≠ b without realizing that if ever a = b, their function will silently and confusingly break. this is why i think it's better to force the user to go "all or nothing" either provide no key, which should break every time, or provide a key, which should never break.

all of this, of course, on top of the new span-based key approach, as spans can't dynamically change like that and shouldn't ever cause this sort of problem.

cAttte commented:
regarding private keys: i think i understand now. a private key is a span+string combo, while a public key is a plain string that universally identifies this state, in the same way that `state` currently works. right?

i think public keys are really bad and shouldn't exist. if you wanna share state between modules, just pass the `state` object around. stringly-typed global state names are not a proper way of communicating between modules.

T0mstone commented:
> regarding private keys: i think i understand now. a private key is a span+string combo (which in my opinion is the ideal solution), while a public key is a plain string that universally identifies this state, in the same way that `state` currently works. right?

Yes, exactly.

> i think public keys are really bad and shouldn't exist. if you wanna share state between modules, just pass the `state` object around. stringly-typed global state names are not a proper way of communicating between modules.

I feel like there might be some situations where it could be useful, but I also can't come up with anything concrete. Either way, it might be good to keep it temporarily (maybe as deprecated if people agree that it's bad) to give users time to update.

cAttte commented:
just realized, `counter()` would have to be updated as well. i can think of 3 approaches:

1. keep the current `key: str or selector` parameter, and just make it optional: reasonable, but removes the natural semblance with `state()` which is bad for the user as (a) it suggests that they are two disparate mechanisms, and (b) it compels them to provide a key, when 99% of the time they won't need to (after this change).
2. split the parameter into a positional but optional `target: selector` parameter, and another named `key: str` parameter: this is very strange, particularly because these two parameters are disjoint; a targeted counter with a key makes no sense (the target _is_ the key, or at least it serves its function).
3. split the signature into two named parameters, `target: selector` and `key: str`: again, the arg parser would have to error when both parameters are specified, but i think this is more intuitive when both params are named, because there is a sort of "symmetry". also, `target: selector` is reminiscent of `outline`'s signature, which i think is another nice symmetry. this is lost when both params are merged into one mere positional parameter.

```typ
1. #let counter(key?)
2. #let counter(target?, key: none)
3. #let counter(target: none, key: none)
```

i think #​3 is the right way to go!

T0mstone commented:
I don't understand your issue with number 1, that seems perfectly reasonable to me.
Also, considering the direction counters will probably go in (see the discussion in #5716), this is even an argument to keep `state.key` around and also use it in counters (Tho maybe it being called `state.key` would be a bit confusing, but I don't have a better idea either). That way a future multi-key counter could look like `counter(heading, state.key("theorem", public: false))` and in particular the key could be created once (`#let k = state.key("theorem", public: false)`) and e.g. passed to both `counter(heading, k, ...).get()` and `counter(k).step()`.

T0mstone commented:
I think at least having a private key type is actually a technical necessity, since `StateUpdateElem` has a public `key` field.
(And no, that can't be made `#[internal]`, I tried, but internal fields can't be used with selectors so `state.get()` would stop working.)

T0mstone commented:
I reworked everything based on the feedback here.
Now `state` is either called as `state(init)` or as `state(key, init)` and the key is always private. I also added a warning in a call like `state(init)` if `init` is a string.

Similarly, counter keys are now also private and you can call `counter()` as a shorthand for `counter("")`.

I'm not sure if I used all the special typst documentation syntax correctly, so some feedback on that would be appreciated.

T0mstone commented:
> Making sure that it's it's not immediately breaking.

The whole point of this PR is already breaking, so if people will have to look into how `state` changed anyway, we might as well go with the new behavior anyway. I even already added a warning for this, so I'm kinda doubtful of whether it's really worth it to roll this out in two phases.

Note that the behavior of `state(key, init)` isn't changed, only `state(arg)` went from `state(arg, none)` to `state("", arg)`.

laurmaedje commented:
I disagree that the point of the PR is to be breaking. The point is to enable anonymous states. It might be that we come to the conclusion that the interface we want is breaking and then we have to gauge the impact and see how to deal with that, but there also certainly possible design that are not breaking.

laurmaedje commented:
I haven't made up my mind yet on how to proceed here. Unfortunately, I can't prioritize this now. Let's revisit this after 0.14.

MrQubo commented:
~~Btw, you can have what is essentially an anonymous state by running something like `head -c256 /dev/urandom | sha256sum` in you shell and using the output as a key. But that's a bit silly, I would like to have this implemented as well.~~
Nvm, this doesn't work...
